src/uct/cuda/cuda_copy/cuda_copy_md.h

```diff
@@ -18,33 +19,59 @@ typedef enum {
     UCT_CUDA_PREF_LOC_LAST
 } uct_cuda_pref_loc_t;
 
+typedef enum {
+    UCT_CUDA_ALLOC_METHOD_DEVICE,
+    UCT_CUDA_ALLOC_METHOD_FABRIC_DEVICE,
+    UCT_CUDA_ALLOC_METHOD_LAST
+} uct_cuda_alloc_method_t;
+
 
 /**
  * @brief cuda_copy MD descriptor
  */
 typedef struct uct_cuda_copy_md {
     struct uct_md               super;           /* Domain info */
+    size_t                      granularity;     /* allocation granularity */
     struct {
         ucs_on_off_auto_value_t alloc_whole_reg; /* force return of allocation
                                                     range even for small bar
                                                     GPUs*/
         double                  max_reg_ratio;
         int                     dmabuf_supported;
+        int                     fabric_supported;
         uct_cuda_pref_loc_t     pref_loc;
+        unsigned                num_alloc_methods;
+        uct_cuda_alloc_method_t alloc_methods[UCT_CUDA_ALLOC_METHOD_LAST];
     } config;
 } uct_cuda_copy_md_t;
 
+typedef UCS_CONFIG_ARRAY_FIELD(uct_cuda_alloc_method_t, methods) uct_cuda_alloc_array_t;
+
 /**
- * gdr copy domain configuration.
+ * cuda_copy MD configuration.
  */
 typedef struct uct_cuda_copy_md_config {
     uct_md_config_t             super;
     ucs_on_off_auto_value_t     alloc_whole_reg;
     double                      max_reg_ratio;
     ucs_ternary_auto_value_t    enable_dmabuf;
+    ucs_ternary_auto_value_t    enable_fabric;
     uct_cuda_pref_loc_t         pref_loc;
+    uct_cuda_alloc_array_t      alloc_methods;
 } uct_cuda_copy_md_config_t;
 
+/**
+ * copy alloc handle.
+ */
+typedef struct uct_cuda_copy_alloc_handle {
+    uct_cuda_alloc_method_t      alloc_type;
```

Comment:
can we remove this field, and check value of generic_handle field to know if it is a VMM allocation?

Reply:
We can query the pointer again and see if allocation VMM type to free. I'll change to that for the time being.